src/src/PluginStructs/P145_data_struct.cpp

```diff
@@ -0,0 +1,598 @@
+@return Rzero as calculated by the autocalibration algorithm
+@note   This is the internally corrected Rzero and will be unaltered when
+        Autocalibration is switched off.
+*/
+float P145_data_struct::getAutoCalibrationValue() const
+{
+  return rzero;
+}
+
+/**************************************************************************/
+/*!
+@brief  Dump data from the internal plugin struct/class structure to the log
+@note   This function is only used for debugging
+*/
+/**************************************************************************/
+void P145_data_struct::dump() const
+{
+  if (loglevelActiveFor(LOG_LEVEL_INFO))
+  {
+#ifdef P145_DEBUG
+    addLog(LOG_LEVEL_INFO, concat(F("MQ-xx: NAME "), sensordef.name));
+    addLog(LOG_LEVEL_INFO, concat(F("MQ-xx: CleanRatio "), sensordef.cleanRatio));
+    addLog(LOG_LEVEL_INFO, concat(F("MQ-xx: Algorithm "), sensordef.alg));
+    addLog(LOG_LEVEL_INFO, concat(F("MQ-xx: PARA "), sensordef.para));
+    addLog(LOG_LEVEL_INFO, concat(F("MQ-xx: PARB "), sensordef.parb));
+    addLog(LOG_LEVEL_INFO, concat(F("MQ-xx: CORA "), sensordef.cora));
+    addLog(LOG_LEVEL_INFO, concat(F("MQ-xx: CORB "), sensordef.corb));
+    addLog(LOG_LEVEL_INFO, concat(F("MQ-xx: CORC "), sensordef.corc));
+    addLog(LOG_LEVEL_INFO, concat(F("MQ-xx: CORD "), sensordef.cord));
+    addLog(LOG_LEVEL_INFO, concat(F("MQ-xx: CORE "), sensordef.core));
+    addLog(LOG_LEVEL_INFO, concat(F("MQ-xx: CORF "), sensordef.corf));
+    addLog(LOG_LEVEL_INFO, concat(F("MQ-xx: CORG "), sensordef.corg));
+    addLog(LOG_LEVEL_INFO, concat(F("MQ-xx: rzero "), rzero));
+    addLog(LOG_LEVEL_INFO, concat(F("MQ-xx: PIN: "), analogPin));
+#endif
+  }
+}
+
+/**************************************************************************/
+/* @brief Provide the sensor name associated with a sensor type
+   @param[in] stype Sensor type (index in sensor table)
+   @return Pointer to sensor name as stored in flash memory
+   @note Returns "invalid" for invalid sensor types
+         This is a static member function
+*/
+/**************************************************************************/
+const __FlashStringHelper * P145_data_struct::getTypeName(int stype)
+{
+  if ((stype < nbrOfTypes) && (stype >= 0))
+    return (const __FlashStringHelper *)&sensorData[stype].name;
+  else
+    return F("invalid");
+}
+
+/**************************************************************************/
+/* @brief Provide the name of the gas measured by a sensor type
+   @param[in] stype Sensor type (index in sensor table)
+   @return Pointer to gas name as stored in flash memory
+   @note Returns "invalid" for invalid sensor types
+         This is a static member function
+*/
+/**************************************************************************/
+const __FlashStringHelper * P145_data_struct::getGasName(int stype)
+{
+  if ((stype < nbrOfTypes) && (stype >= 0))
+    return (const __FlashStringHelper *)&sensorData[stype].gas;
```

Comment:
This really is tricky.
Either the type in that struct should then be changed into `const __FlashStringHelper *` instead of `char[8]` or just return type `String`  for this function.
I think casting a char array to a flash string is extremely tricky as there is no guarantee how other functions may try to access this data.
What if there is some check somewhere in the SDK code to check if the address is in the expected range.
No, please change this as it will for sure cause issues and I have no idea how bad those will be.

Reply:
Yeah, I found this somewhere when I was in doubt about how to treat a char[] in PROGMEM. I start to like the magic of the String class. But as an old school embedded C programmer I always tried to keep the number of constructors/destructors in expressions low. But returning a String object makes it much more safe and readable.

Reply:
You can return of type flashstring, but then the members in the PROGMEM struct should also be of this type.
You can interpret flash strings as char pointers and access them using the standard functions with the `_P` postfix (e.g. `memcpy_P`) but not the other way around.
You cannot cast something which isn't a flash string into a flashstring type.
Well you can and the compiler doesn't think it's wrong, but you don't know how other functions will try to access the content of something that's marked as a flash string.